src/client/Client.cc

```diff
@@ -5784,17 +5761,58 @@ void Client::flush_mdlog(MetaSession *session)
 }
 
 
-void Client::_unmount()
+void Client::_abort_mds_sessions(int err)
+{
+  for (auto p = mds_requests.begin(); p != mds_requests.end(); ) {
+    auto req = p->second;
+    ++p;
+    // unsafe requests will be removed during close session below.
+    if (req->got_unsafe)
+      continue;
+
+    req->abort(err);
+    if (req->caller_cond) {
+      req->kick = true;
+      req->caller_cond->Signal();
+    }
+  }
+
+  // Progress aborts on any requests that were on this waitlist.
+  // Any requests that were on a waiting_for_open session waitlist
+  // will get kicked during close session below.
+  signal_cond_list(waiting_for_mdsmap);
+
+  // Force-close all sessions
+  while(!mds_sessions.empty()) {
+    auto& session = mds_sessions.begin()->second;
+    _closed_mds_session(&session);
+  }
+}
+
+void Client::_unmount(bool abort)
 {
   if (unmounting)
     return;
 
-  ldout(cct, 2) << "unmounting" << dendl;
+  if (abort || blacklisted) {
+    ldout(cct, 2) << "unmounting (" << (abort ? "abort)" : "blacklisted)") << dendl;
+  } else {
+    ldout(cct, 2) << "unmounting" << dendl;
+  }
   unmounting = true;
 
   deleg_timeout = 0;
 
-  flush_mdlog_sync(); // flush the mdlog for pending requests, if any
+  if (abort) {
+    // Abort all mds sessions
+    _abort_mds_sessions(-ENOTCONN);
+
+    objecter->op_cancel_writes(-ENOTCONN);
```

Comment:
In principle, what we want to do here is flush any cached dirty data we have out to the OSDs while shutting down. The thing we want to preserve is existing stateful objects (opens, locks, delegations, etc). For instance, if we have Fb caps, and have some data buffered up, we should probably flush that to the OSD before shutting down, rather than cancelling those writes (and trashing the cache).

Note that NFS clients should still attempt to replay writes that didn't get committed when the server comes back, so it shouldn't be fatal if we can't do that, but it might be nicer overall.

Reply:
how about calling sync_fs before calling abort_conn?

Reply:
That's probably fine, and good as a matter of course anyway. I'll plan to do that. I guess I was just thinking about the mdsc and osdcs separately, and that abort_conn would abort the former but not the latter. Maybe it's simpler for users to keep it all together though like you suggest.

Reply:
The problem is that if we keep data IOs, we need to keep inodes with dirty data in the cache.

Reply:
Ok, sounds good.